frontend/src/metabase/query_builder/components/view/NativeQueryPreviewSidebar/NativeQueryPreviewSidebar.tsx

```diff
@@ -0,0 +1,124 @@
+import { t } from "ttag";
+
+import { useGetNativeDatasetQuery } from "metabase/api";
+import { DelayedLoadingSpinner } from "metabase/common/components/EntityPicker/components/LoadingSpinner";
+import { color } from "metabase/lib/colors";
+import { formatNativeQuery, getEngineNativeType } from "metabase/lib/engine";
+import { useDispatch, useSelector } from "metabase/lib/redux";
+import { checkNotNull } from "metabase/lib/types";
+import { updateQuestion, setUIControls } from "metabase/query_builder/actions";
+import { NativeQueryEditorRoot } from "metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.styled";
+import { getQuestion } from "metabase/query_builder/selectors";
+import { Box, Button, Flex, Icon, rem } from "metabase/ui";
+import * as Lib from "metabase-lib";
+
+import SB from "./NativeQueryPreviewSidebar.module.css";
+import { createDatasetQuery } from "./utils";
+
+const TITLE = {
+  sql: t`SQL for this question`,
+  json: t`Native query for this question`,
+};
+
+const BUTTON_TITLE = {
+  sql: t`Convert this question to SQL`,
+  json: t`Convert this question to a native query`,
+};
+
+export const NativeQueryPreviewSidebar = (): JSX.Element => {
+  const dispatch = useDispatch();
+  const question = checkNotNull(useSelector(getQuestion));
+
+  const engine = question.database()?.engine;
+  const engineType = getEngineNativeType(engine);
+
+  const sourceQuery = question.query();
+  const canRun = Lib.canRun(sourceQuery);
+  const payload = Lib.toLegacyQuery(sourceQuery);
+  const { data, error, isLoading } = useGetNativeDatasetQuery(payload);
+
+  const showError = canRun && error;
+  const showQuery = canRun && !error;
+  const showEmptySidebar = !canRun;
+
+  const formattedQuery = formatNativeQuery(data?.query, engine);
+
+  const handleConvertClick = useCallback(() => {
+    if (!formattedQuery) {
+      return;
+    }
+
+    const newDatasetQuery = createDatasetQuery(formattedQuery, question);
+    const newQuestion = question.setDatasetQuery(newDatasetQuery);
+
+    dispatch(updateQuestion(newQuestion, { shouldUpdateUrl: true, run: true }));
+    dispatch(setUIControls({ isNativeEditorOpen: true }));
+  }, [question, dispatch, formattedQuery]);
+
+  const getErrorMessage = (error: unknown) =>
+    typeof error === "string" ? error : undefined;
+
+  const borderStyle = `1px solid ${color("border")}`;
+
+  return (
+    <Flex
+      data-testid="native-query-preview-sidebar"
+      className={cx(SB.container)}
+      role="complementary"
+      direction="column"
+      style={{ borderLeft: borderStyle }}
+    >
+      <Box
+        component="header"
+        c={color("text-dark")}
+        fz={rem(20)}
+        lh={rem(24)}
+        fw="bold"
+        ta="start"
+        p="1.5rem"
+      >
+        {TITLE[engineType]}
+      </Box>
+      <Box
+        style={{ flex: 1, borderTop: borderStyle, borderBottom: borderStyle }}
+      >
+        {isLoading && <DelayedLoadingSpinner delay={1000} />}
```

Comment:
> We don’t need to give the user any visual feedback that the SQL is live-updating unless the API call takes more than a second

This does not work. We should use `isFetching` instead of `isLoading`.
And when we do, things get shifted around.


https://github.com/metabase/metabase/assets/6830683/cddea00e-4a45-42d5-91d1-82bf7bd2c42c